cmp/grid/GridStateModel.js

```diff
@@ -0,0 +1,196 @@
+    readState(stateKey) {
+        return XH.localStorageService.get(stateKey, {});
+    }
+
+    saveState(stateKey, state) {
+        XH.localStorageService.set(stateKey, state);
+    }
+
+    resetState(stateKey) {
+        XH.localStorageService.remove(stateKey);
+    }
+
+    loadState(state) {
+        this.state = cloneDeep(state || this.readState(this.getStateKey()) || {});
+        this.updateGridColumns();
+        this.updateGridSort();
+    }
+
+    resetStateAsync() {
+        const defaultState = this.defaultState;
+
+        if (!defaultState) resolve();
+
+        return start(() => {
+            this._resetting = true;
+            this.loadState(defaultState);
+        }).then(() => {
+            this.resetState(this.getStateKey());
+        }).finally(() =>{
+            this._resetting = false;
+        });
+    }
+
+    //--------------------------
+    // Columns
+    //--------------------------
+    columnReaction() {
+        const {parent} = this;
+        return {
+            track: () => parent.columns,
+            run: () => {
+                this.state.columns = this.getColumnState();
+                this.saveStateChange();
+            }
+        };
+    }
+
+    getColumnState() {
+        if (!this.trackColumns) return undefined;
+
+        const columns = this.parent.columns;
+
+        return columns.map(it => {
+            return {
+                xhId: it.xhId,
+                hide: it.hide,
+                width: it.width
+            };
+        });
+    }
+
+    updateGridColumns() {
+        const {parent, state} = this,
+            cols = parent.cloneColumns(),
+            newColumns = [],
+            foundColumns = [];
+
+        if (this.trackColumns && state.columns) {
+            state.columns.forEach(colState => {
+                const col = find(cols, {xhId: colState.xhId});
+                if (!col) return;
+
+                col.hide = colState.hide;
+                newColumns.push(col);
+                foundColumns.push(col);
+            });
+
+            cols.forEach((col, idx) => {
+                if (!find(foundColumns, {xhId: col.xhId})) {
+                    newColumns.splice(idx, 0, col);
+                }
+            });
+
+            parent.setColumns(newColumns);
+        }
+
+    }
+
+    //--------------------------
+    // Sort
+    //--------------------------
+    sortReaction() {
+        const {parent} = this;
+        return {
+            track: () => parent.sortBy,
+            run: () => {
+                this.state.sortBy = parent.sortBy;
+                this.saveStateChange();
+            }
+        };
+    }
+
+    updateGridSort() {
+        const sortBy = this.state.sortBy;
+        if (sortBy) this.parent.setSortBy(sortBy);
+    }
+
+    //--------------------------
+    // Helper
+    //--------------------------
+    saveStateChange = debounce(function() {
+        this.saveState(this.getStateKey(), this.state);
+    }, 5 * SECONDS);
```

Comment:
In Hoist-Sencha we check for the state object and whether or not we are resetting. I think we should always have a state object (even if it's empty).

As far as resetting, I think if we keep the asynchronous state reset we should check for it, but if we make that operation synchronous it's probably unnecessary.